Refactored `VMenu::ShowMenu` and its vicinity.

MKadaner opened the pull request:
## Summary
Refactoring in preparation for #711.

## References
- #711

## Checklist
- [X] I have followed the [contributing guidelines](https://github.com/FarGroup/FarManager/blob/master/CONTRIBUTING.md).
- [ ] I have discussed this with project maintainers: <!-- add a link to the corresponding issue / discussion / forum topic here --> <br/>
If not checked, I accept that this work might be rejected in favor of a different great big ineffable plan.<br/>

## Details
Sorry, it is large, but I believe overall the code is cleaner and easier to understand.

CodeFactor is complaining, and probably rightfully so. On the positive side, one issue is fixed.

MKadaner commented:
CodeFactor is complaining, and probably rightfully so. On the positive side, one issue is fixed.

alabuzhev commented on far/vmenu.hpp:
```diff
@@ -53,25 +53,25 @@ THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 // Цветовые атрибуты - индексы в массиве цветов
-enum vmenu_colors
+enum class vmenu_colors
```
Do we actually need this in .hpp?

MKadaner replied:
Down below, we have

```C++
FarColor Colors[std::to_underlying(vmenu_colors::COUNT)]{};
```

Is there a way to get the number of Far colors (15) without exposing this definition?

alabuzhev replied:
vector maybe?

MKadaner replied:
Did it another way. Please see if you like it.

alabuzhev replied:
It's ok, as long as we assert the equality.

MKadaner replied:
We do. Resolving this one.

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -1957,32 +2054,34 @@ void VMenu::DisplayObject()
 	if (!CheckFlags(VMENU_DISABLEDRAWBACKGROUND) && !CheckFlags(VMENU_LISTBOX))
 	{
-		// BUGBUG, dead code
+		// BUGBUG, dead code -- 2023-07-08 MZK: Is it though? I've got here when pressed hotkey of "Select search area" combobox on "find file" dialog.
```
TBH I don't remember the reason of the comment.

Maybe it applies only to the `!CheckFlags(VMENU_DISABLEDRAWBACKGROUND)` part of the condition, which is always true, since nothing sets the flag anywhere (perhaps we should remove it).
Or maybe I didn't notice the `!` and thought that the whole branch is unreachable.

MKadaner replied:
Interesting. I adjusted the comment.

If this this flag is indeed effectively unused, can we repurpose it? I need a new flag for enabling vertical alignment of annotations feature. It should be enabled only for specific lists, not for regular vertical menus, for example. Unfortunately, `VMENU_FLAGS` utilize all 32 bits, so I was not sure how to fit another one.

alabuzhev replied:
Can't see why not (with a new name obviously).

MKadaner replied:
Thanks. I thought about plugin interface which might be affected. I am always uncertain about plugin API.

Leaving it for now, and will get back to it when implementing vertical alignment of annotations.

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -1834,15 +1877,69 @@ bool VMenu::ShiftAllItemsShowPos(int Shift)
 	return NeedRedraw;
 }
 
+struct item_layout
```
Wow, it's finally structured.
Thank you.

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -2129,253 +2331,139 @@ void VMenu::ShowMenu(bool IsParent)
-					MenuItemForDisplay.resize(MaxLineWidth);
-					ShowRightScroller = true;
-				}
+	separator.at(NamePos - 1) = L' ';
```
Why `at`, do we expect exceptions?

MKadaner replied:
Preexisting code, but of course an exception is not expected. Fixed.

alabuzhev replied:
If you're about `VMenu::at` - that's different.
VMenu is usually accessed via a (smart) pointer and writing `menu->at(idx)` is way more natural than `(*menu)[idx]`.

MKadaner replied:
Ah, sorry, it was I who introduced it here. I actually merged my changed from a different branch where I could not be so sure that the `separator` string is of correct size. It should not happen in this context.

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -2129,253 +2331,139 @@ void VMenu::ShowMenu(bool IsParent)
+}
 
-				strMenuLine += MenuItemForDisplay;
+void VMenu::DrawRegularItem(const MenuItemEx& CurItem, const item_layout& Layout, const int Y, std::vector<int>& HighlightMarkup, const string& BlankLine) const
```
Any reason BlankLine can't be const string_view?

MKadaner replied:
No reasons. Fixed. Thank you. Omitted `const` because `string_view` is effectively const anyway. Let me know if you agree.

alabuzhev replied:
const is meaningful even on string_view: it means that you can't (accidentally) change the view itself, e.g. by calling remove_suffix or reassigning it altogether.

MKadaner replied:
OK, I should've known your tastes better. 🙂

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -2129,253 +2331,139 @@ void VMenu::ShowMenu(bool IsParent)
-					Text(L'«');
-				}
+	SetColor(ColorIndices.Normal);
+	Text(string_view{ BlankLine.begin(), BlankLine.begin() + (TextBegin + TextWidth - WhereX()) });
```
BlankLine.substr(...) maybe?
(which should be cheap if BlankLine is string_view)

MKadaner replied:
Sure!

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -3048,4 +3129,72 @@ TEST_CASE("find.nearest.selectable.item")
+
+TEST_CASE("markup.slice.boundaries")
+{
+	static struct test_data
```
Can it be const?

MKadaner replied:
Absolutely!

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -3048,4 +3129,72 @@ TEST_CASE("find.nearest.selectable.item")
+	static struct test_data
+	{
+		std::pair<int, int> Segment;
+		std::vector<std::pair<int, int>> Slices;
```
Do we need vectors here?
The data is static, so spans, initializer_lists or arrays should probably do.

MKadaner replied:
Ah, `initializer_list` works with `const`, but not with `constexpr` here. I originally played with `constexpr`, could not make it work, and gave up. Went with `vector`s and lost `const`. :( Thank you for calling this out. fixed.

alabuzhev reviewed (commented):
It looks way too complex for a thorough review :)
I trust you spent quite some time debugging it and know what you're doing, so the comments will likely be only about technical details & style.

MKadaner commented on far/vmenu.hpp:
```diff
@@ -318,6 +331,8 @@ class VMenu final: public Modal
 	void UpdateSelectPos();
 	void EnableFilter(bool Enable);
 
+	[[nodiscard]] FarColor GetColor(vmenu_colors ColorIndex) const noexcept { return Colors[std::to_underlying(ColorIndex)]; }
```
Nope! Thanks for pointing out.

MKadaner commented on far/vmenu.cpp:
```diff
@@ -172,6 +171,47 @@ int find_nearest(std::ranges::contiguous_range auto const& Range, const int Pos,
 			: FindPos(drop(Pos), take(Pos) | reverse));
 }
 
+template <std::integral T>
```
Well, we can specialize them for the `std::pair<int, int>` which is the only one required, and for `std::vector<int>& Markup`, though I am not sure it makes things really easier to understand. What we cannot do is to avoid generic `std::ranges::input_range auto const& Slices` because we actually call this function with different view types. I will apply the changes.

MKadaner commented on far/vmenu.cpp:
```diff
@@ -172,6 +171,47 @@ int find_nearest(std::ranges::contiguous_range auto const& Range, const int Pos,
+		std::swap(A, B);
+
+	if (A.second <= B.first)
+		return { {}, {} };
```
Good point! Thanks.

MKadaner commented on far/vmenu.cpp:
```diff
@@ -172,6 +171,47 @@ int find_nearest(std::ranges::contiguous_range auto const& Range, const int Pos,
+{
+	assert(Segment.first < Segment.second);
+
+	for (const auto Slice : Slices)
```
`Slices` is a collection of `std::pair<int, int>`, so there is not much use in the extra indirection.

I tried to restrict the type of the input range element here, but all my attempts look quite ugly. They could be acceptable in a general-purpose library but not in this one-off context. If you can suggest something you would accept here, I'll be glad to use it and learn.

alabuzhev replied:
I mentioned it mostly because static analyzers don't like by-value-for-loops at all and strongly suggest it's an error, without paying much attention to the underlying type.
So to see fewer bogus warnings it might be better to go with the flow.

MKadaner replied:
OK. Hopefully, compiler will optimize it out anyway.

MKadaner commented:
> [T]he comments will likely be only about technical details & style.

I addressed all comments so far. Could you please check them and resolve whatever looks good?

> I trust you spent quite some time debugging it and know what you're doing...

Well, yes, but... Thorough testing is difficult and time consuming.

@rohitab, if by any chance you can spend some time testing this change, I would hugely appreciate it. The change affects all kinds of vertical lists: menus, histories, Find All, comboboxes... The bugs are expected with vertical layout (irregular movement of selection or items, unexpected empty lines, etc.) and with drawing list borders.

I will of course test more during the long weekend. Just now, I noticed one glitch. Find File -> Code Page Combobox may appear something like this (note the mangled left border; if there is no scrollbar, the right border should look similar):

```
Using code page:    
Automatic detection 
┌───────────────────
├  1201 │ UTF-16 (Bi
├───────┼───────────
├  37   │ IBM EBCDIC
├  500  │ IBM EBCDIC
├  708  │ Arabic - A
├  720  │ Arabic - T
├  737  │ OEM - Gree
│  775  │ OEM - Balt
└───────────────────
```

I guess I know how to fix this one (even two ways). Will do during the weekend as well.

MKadaner commented:
Fixed everything, including the bug with vertical borders.

Not squashing yet to make verification of the comments easier. Also, I want to test a bit more.

Please let me know if everything is OK.

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -172,6 +171,45 @@ int find_nearest(std::ranges::contiguous_range auto const& Range, const int Pos,
+	assert(B.first < B.second);
+
+	if (B.first < A.first)
+		std::swap(A, B);
```
Let's use ranges::swap for consistency.
Same below.

MKadaner replied:
OK. Sorry, below where?

alabuzhev replied:
Yes

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -1834,15 +1875,69 @@ bool VMenu::ShiftAllItemsShowPos(int Shift)
+			Text = { Left, Right + 1 - Left };
+	}
+
+	[[nodiscard]] static bool NeedBox(int BoxType) noexcept { return BoxType != NO_BOX; }
```
Should these static methods be in the struct?
If yes (e.g. to and highlight that they're about layout) it's fine, but it would be better to keep the same case style: `item_layout::GetServiceAreaSize` looks a bit eclectic.

MKadaner replied:
Two of them, `NeedSubMenu` and `NeedScrollbar`, need access to `VMenu` privates, as well as `item_layout` constructor. For this purpose, `item_layout` is made a friend. I'll rename them to snake_case.

alabuzhev commented on far/vmenu.cpp:
```diff
@@ -2129,253 +2365,139 @@ void VMenu::ShowMenu(bool IsParent)
+	for (const auto SliceEnd : HighlightMarkup)
+	{
+		SetColor(Colors, CurColorIndex);
+		Text(string_view{ ItemTextToDisplay.begin() + CurPos, ItemTextToDisplay.begin() + SliceEnd });
```
`string_view(ItemTextToDisplay).substr(CurPos, SliceEnd - CurPos)` maybe?

MKadaner replied:
I thought about it. No real preference. Let's do it your way.

alabuzhev commented:
The change looks good, although I haven't tested it.
 
> Thorough testing is difficult and time consuming.

Indeed. We're not in a hurry, so if you'd like to have another look or play more with it, please take your time, we can get back to it next year :)

MKadaner commented:
Thank you for the approval. I have been playing with it today for an hour or so and could not find any new issues. Let's call it a day.

I rebased and squashed. Please merge.

alabuzhev commented:
Ok. Thank you.

rohitab commented:
@MKadaner sorry I wasn't able to respond sooner. I am going to upgrade to this version and will let you know if I come across any issues.

MKadaner commented:
@rohitab, no problem at all. Your feedback is welcome at any time. When you upgrade, please pick the latest version. I pushed one more refactoring in the same area, drawing of the vertical lists to be more specific. The behavior should not be affected.
